### ISSUE TITLE
@EqualsAndHashcode violates symmetricity rule for polymorphic classes if not all classes use Lombok

### ISSUE THREAD

**Issue by ewirch** [rzwitserloot/lombok#1406]
The JLS requires symmetricity for `equals()` implementations:

>  for any non-null reference values x and y, x.equals(y) should return true if and only if y.equals(x) returns true.

Consider this code:

    @Test
    public void testEqualsAndHashCode() {
    	final X x = new X("1", "2");
    	final Y y = new Y("1", "2", "3");
    	assertFalse(y.equals(x));
    	assertFalse(x.equals(y));
    }
    
    @EqualsAndHashCode
    @RequiredArgsConstructor
    public static class X {
    	private final String property1;
    	private final String property2;
    }
    
    public static class Y extends X {
    	private final String property3;
    
    	public Y(final String property1, final String property2, final String property3) {
    		super(property1, property2);
    		this.property3 = property3;
    	}
    
    	@Override
    	public boolean equals(final Object o) {
    		if (this == o) return true;
    		if (o == null || getClass() != o.getClass()) return false;
    		if (!super.equals(o)) return false;
    
    		final Y y = (Y) o;
    
    		if (property3 != null ? !property3.equals(y.property3) : y.property3 != null) return false;
    
    		return true;
    	}
    }

`y.quals(x)` does not return `true`, so `x.equals(y)` is not allowed to return true. But it does return true!

`canEqual()` is a  mere workaround not a solution to this problem. It is prone to heterogeneous `equals()` implementations. Default implementation of `@EqualsAndHashCode` should use `getClass() == other.getClass()`.

**Comment by irobertson** [comment]
The suggested change would fix one problem, but introduce another. There are plenty of valid use cases for instances of different classes being equal to each other.

While searching for a more inclusive solution, it would be good to amend the documentation of EqualsAndHashCode to indicate that, in addition to overriding canEqual if writing your own equals method, you should also consider overriding it if you plan on subclassing with classes which cannot be equal to instances of the original class.

**Comment by victorwss** [comment]
@irobertson I think that making objects of different classes being equals is a bad idea and a bad practice. It is almost always simply buggy. If it is done intentionally, this is a sloppy programming.

It would be nice if lombok generates an `equals` methods that never makes the object equals to something else from another class. If the user wants it to be considered equals to something else from another class, then he/she is free to write the `equals` method for that by hand and take the burden and the risks of that decision for himself/herself.

**Comment by Maaartinus** [comment]
> `y.quals(x)` does not return true, so `x.equals(y)` is not allowed to return true. But it does return true!

@ewirch Wrong! `x.equals(y)` does not return false, so `y.equals(x)` is not allowed to return false. Does a superclass have to obey the contract of a subclass or the other way round?

> `canEqual()` is a mere workaround not a solution to this problem.

The sole problem is the broken `Y#equals`. `X` was perfectly fine before that. The purpose of `canEqual()` is not the solve the non-problem, but rather to allow both behaviors: allow or disallow subclass equality.

> Default implementation of `@EqualsAndHashCode` should use `getClass() == other.getClass()`.

I might agree with "default implementation of `canEqual` should use `getClass() == other.getClass()`".

---

> While searching for a more inclusive solution

@irobertson Maybe adding a *warning on non-final classes* which define no `canEqual`? To keep it concise, adding a `subclassesCanEqual` parameter as a shortcut? Final classes are fine and for the others we have
- `subclassesCanEqual=false`, no `canEqual` defined → 
    generate `canEqual` testing `getClass() == other.getClass()`
- `subclassesCanEqual=true`, no `canEqual` defined →
    generate `canEqual` like before
- no `subclassesCanEqual` set, but `canEqual` defined →
   let's hope you know what you're doing
- anything else →
  a warning

---

> I think that making objects of different classes being equals is a bad idea

@victorwss Tell everyone using dynamic proxies (e.g., Hibernate users).

> ....

If the user :rofl: *does not want* :rofl: it to be considered equals to something else from another class, then he/she is free to write the equals method for that by hand and take the burden and the risks of that decision for himself/herself.  

There were quite a few discussions about `this.class == that.class` vs. `that instanceof ThisClass` and IIRC both camps were about the same size. AFAIK `canEqual` is the only solution satisfying both.
